### DISCUSSION TITLE
conversion from 'const Ice::Long' to 'long' requires a narrowing conversion

### DISCUSSION THREAD

**Discussion by domcir** [zeroc-ice/ice#3173] [Q&A]
Hi
I'm trying to use ice on windows 64x and I get the conversion issues for long type.
I'm using Ice 3.7.6.
The same implmenetation works well on linux so not sure what might be wrong.
I managed to find the below definition where I would like to use long instead of __int64 but not sure how to make it happen.
![image](https://github.com/user-attachments/assets/c0b1c2d3-8b21-41cf-ac06-5d99d09d5bf2)
Thanks
Dominik

**Comment by pepone** [comment]
Hi Dominik,

What C++ compiler are you using? Can you show us the code and the actual compiler warnings?

Cheers,
Jose

**Reply by domcir** [reply]
VS14.2
![image](https://github.com/user-attachments/assets/435e174b-fdaf-441d-a578-6c4784833842)
I'm not using ICE_CPP11_MAPPING.

### Example code (simplified as much as I could)
**1. ice object description**

 class Instrument {
                long                                        m_id;
                string                                      m_description;
                string                                      m_annotations;
            };
**2. generated class**

    class ICE_CLASS(platform_all_ice) Instrument : public ::Ice::ValueHelper<Instrument, ::Ice::Value>
    {
    public:

    ICE_MEMBER(platform_all_ice) virtual ~Instrument();

    Instrument() = default;

    Instrument(const Instrument&) = default;
    Instrument(Instrument&&) = default;
    Instrument& operator=(const Instrument&) = default;
    Instrument& operator=(Instrument&&) = default;

    /**
     * One-shot constructor to initialize all data members.
     */
    Instrument(long long int m_id, const ::std::string& m_description, const ::std::string& m_annotations) :
        m_id(m_id),
        m_description(m_description),
        m_annotations(m_annotations),
    {
    }

    /**
     * Obtains a tuple containing all of the value's data members.
     * @return The data members in a tuple.
     */
    std::tuple<const long long int&, const ::std::string&, const ::std::string&> ice_tuple() const
    {
        return std::tie(m_id, m_description, m_annotations);
    }

    /**
     * Obtains the Slice type ID of this value.
     * @return The fully-scoped type ID.
     */
    ICE_MEMBER(platform_all_ice) static const ::std::string& ice_staticId();

    long long int m_id;
    ::std::string m_description;
    ::std::string m_annotations;
    };

**3. usage**

    Instruemnt i();
    long i = i.m_id   ///<-- error only win, ok with linux

Severity	Code	Description
Error	C2398	Element '1': conversion from '__int64' to 'long' requires a narrowing conversion

**Comment by pepone** [comment]
Slice `long` maps to a 64 bit number, the size of C++ long is platform dependent. When using the C++98 mapping, you should use `Ice::Long`, which maps to the appropriate type on each platform. For C++11 mapping, long long is used, as it is guaranteed to have at least 64 bits.

The generated code for C++ 98 mapping is:

```
    ::Ice::Long m_id;  
    ::std::string m_description;
    ::std::string m_annotations;   
```

The code you posted corresponds to the C++11 mapping.

**Reply by domcir** [reply]
Sorry actually both C++98 and C++11 classes are generated, I just copied wrong class.
The solution is to use ::Ice::Long.
Thanks pepone for you help.
